frontend/lib/resolution/VarScopeVisitor.cpp

```diff
@@ -519,6 +571,53 @@ void VarScopeVisitor::exit(const Conditional* cond, RV& rv) {
   exitAst(cond);
 }
 
+bool VarScopeVisitor::enter(const Select* sel, RV& rv) {
+  enterAst(sel);
+  enterScope(sel, rv);
+  
+  //traverse the not paramFalse children
+  for(int i = 0; i < sel->numWhenStmts(); i++) {
+    auto whenAst = sel->whenStmt(i);
+    bool anyParamTrue = false;
+    bool allParamFalse = !whenAst->isOtherwise();
+    for(auto caseExpr : whenAst->caseExprs()) {
+      auto res = rv.byAst(caseExpr);
+      anyParamTrue = anyParamTrue || res.type().isParamTrue();
+      allParamFalse = allParamFalse && res.type().isParamFalse();
+    }
+    if (!allParamFalse) {
+      whenAst->traverse(rv);
+      currentWhenFrame(i)->paramTrue = anyParamTrue;
+    } else {
+      CHPL_ASSERT(currentWhenFrame(i) == nullptr);
+    }
+    if (anyParamTrue) {
+      break;
+    }
+  }
+
+  //if all branches are param false, the otherwise branch is param true
+  if (sel->hasOtherwise()) {
+    bool allParamFalse = true;
+    for(int i = 0; i < sel->numWhenStmts(); i++) {
+      auto whenFrame = currentWhenFrame(i);
+      if (!whenFrame || sel->whenStmt(i)->isOtherwise()) continue;
+      allParamFalse = false;
+    }
+    if (allParamFalse) {
+      auto otherwiseFrame = currentWhenFrame(sel->numWhenStmts() - 1);
+      otherwiseFrame->paramTrue = true;
+    }
+  }
```

Comment:
The following three things are true right above the `if (anyParamTrue)` statement here if the current node is an `otherwise`:

1. It's the last node (via your post-parse checks constraint).
2. If any of the previous branches was `param true`, you would not have gotten here, because of the `break`. Thus, either a condition was non-param, or all previous conditions were false.
3. You have to track if any prior conditions are non-param to avoid mis-setting `paramTrue` (see my other comment).

Thus, `if (!anyNonParam)`, then `allParamFalse` up until now, and the `otherwise` is `paramTrue`. I.e.:

```C++
    if (whenAst->isOtherwise()) {
      currentWhenFrame(i)->paramTrue = !anyNonParam;
    } else if (anyParamTrue) {
      break;
    }
```
And you don't need any of the code from the second `sel->hasOtherwise()` condition.

Reply:
I think its enough to track if all the `when` blocks are paramFalse rather than if there's been any non params at all, as if there was a paramTrue, we would have broken out of the loop. 

So with a variable renaming, I think the following at the end of the loop over the `when` statements will do the trick:
```
if (whenAst->isOtherwise()) {
  currentWhenFrame(i)->paramTrue = allWhensParamFalse;
}
allWhensParamFalse &= allCasesParamFalse;
```
where `allCasesParamFalse` and `anyCaseParamTrue` are calculated for each `when` node, and `allWhensParamFalse` accumulates for the entire `select`.